File: Modules/module_sample/src/main/java/com/example/william/my/module/sample/model/ArticleViewModel.java
package com.example.william.my.module.sample.model;

import androidx.arch.core.util.Function;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import com.example.william.my.core.network.retrofit.response.RetrofitResponse;
import com.example.william.my.module.bean.ArticleDataBean;
import com.example.william.my.module.bean.ArticleDetailBean;
import com.example.william.my.module.sample.repo.ArticleRepository;

import java.util.List;

public class ArticleViewModel extends ViewModel {

    private int mPage;

    private final ArticleRepository articleRepository;

    private final MutableLiveData<Integer> mutableArticleList;

    private final LiveData<RetrofitResponse<List<ArticleDetailBean>>> mObservableArticleList;

    private final MutableLiveData<Integer> mutableArticle;

    private final LiveData<RetrofitResponse<ArticleDataBean>> mObservableArticle;

    public ArticleViewModel() {

        articleRepository = ArticleRepository.getInstance();

        mutableArticleList = new MutableLiveData<>();

        mObservableArticleList = Transformations.switchMap(mutableArticleList, new Function<Integer, LiveData<RetrofitResponse<List<ArticleDetailBean>>>>() {
            @Override
            public LiveData<RetrofitResponse<List<ArticleDetailBean>>> apply(Integer input) {
                return articleRepository.getArticleList(input);
            }
        });

        mutableArticle = new MutableLiveData<>();

        mObservableArticle = Transformations.switchMap(mutableArticle, new Function<Integer, LiveData<RetrofitResponse<ArticleDataBean>>>() {
            @Override
            public LiveData<RetrofitResponse<ArticleDataBean>> apply(Integer input) {
                return articleRepository.getArticle(input);
            }
        });
    }

    /**
     * Expose the LiveData Comments query so the UI can observe it.
     */
    public LiveData<RetrofitResponse<List<ArticleDetailBean>>> getArticleList() {
        return mObservableArticleList;
    }

    public void onRefreshArticleList() {
        mPage = 0;
        mutableArticleList.postValue(mPage);
    }

    public void onLoadMoreArticleList() {
        mPage++;
        mutableArticleList.postValue(mPage);
    }

    public boolean isFirst() {
        return mPage == 0;
    }

    /**
     * Expose the LiveData Comments query so the UI can observe it.
     */
    public LiveData<RetrofitResponse<ArticleDataBean>> getArticle() {
        return mObservableArticle;
    }

    public void onRefreshArticle() {
        mPage = 0;
        mutableArticle.postValue(mPage);
    }

    public void onLoadMoreArticle() {
        mPage++;
        mutableArticle.postValue(mPage);
    }
}

File: modules/module_sample/src/main/kotlin/com/example/william/my/module/sample/frame/model/ArticleViewModel.kt
package com.example.william.my.module.sample.frame.model

import androidx.lifecycle.*
import com.example.william.my.bean.data.ArticleDataBean
import com.example.william.my.bean.repo.ArticleRepository
import com.example.william.my.core.retrofit.response.RetrofitResponse

class ArticleViewModel(private val articleSource: ArticleRepository) : ViewModel() {

    private val mMutableLiveData = MutableLiveData<Int>()

    val article: LiveData<RetrofitResponse<ArticleDataBean>>
        get() = Transformations.switchMap(mMutableLiveData) { input ->
            articleSource.loadArticle(input)
        }

    fun queryArticle(page: Int) {
        mMutableLiveData.postValue(page)
    }
}

/**
 * 自定义实例，多参构造
 * Factory for [ArticleViewModel].
 */
object ArticleVMFactory : ViewModelProvider.Factory {

    private val dataSource = ArticleRepository.getInstance()

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        @Suppress("UNCHECKED_CAST")
        return ArticleViewModel(dataSource) as T
    }
}
